jupyterhub/app.py

```diff
@@ -2013,12 +2013,19 @@ def _user_summary(user):
         # parallelize checks for running Spawners
         # run query on extant Server objects
         # so this is O(running servers) not O(total users)
+        # Server objects can be associated with either a Spawner or a Service,
+        # we are only interested in the ones associated with a Spawner
         check_futures = []
         for orm_server in db.query(orm.Server):
-            orm_spawners = orm_server.spawner
-            if not orm_spawners:
+            orm_spawner = orm_server.spawner
+            if not orm_spawner:
+                # sanity check for orphaned Server rows
+                # this shouldn't happen if we've got our sqlachemy right
+                if not orm_server.service:
+                    self.log.warning("deleting orphaned server %s", orm_server)
```

Comment:
This is a condition that shouldn't happen (`not server.spawner and not server.service`), but is being cleaned up anyway because I've been known to make mistakes that leave orphaned things in the database.

If the ORM is set up correctly, deleting Spawner and Service objects should mean that associated Server objects also get deleted. There should be no Server objects not associated with Spawners or Services.

Since we are now already iterating over all the Server objects, this bit here is a check to validate that assertion, and delete any Server objects that may still exist without valid relationships (e.g. a Spawner was deleted but its Server was orphaned instead of being deleted at the same time). Basically, it's cleaning up after hypothetical bugs in past versions of the ORM code.

If everything is correct (and always has been correct), this condition will never be met.